File: oxidize-golang/core/autotune/detect.go
// Package autotune mirrors oxidize_core::autotune — hardware detection and
// rule-based inference tuning plans.
package autotune

import (
	"os"
	"runtime"
	"strconv"
	"strings"

	"github.com/Zapdev-labs/oxidize/golang/core/gpucluster"
	"github.com/Zapdev-labs/oxidize/golang/core/simd"
)

// OsKind identifies the host operating system.
type OsKind int

const (
	OsLinux OsKind = iota
	OsMacos
	OsWindows
	OsOther
)

func (o OsKind) String() string {
	switch o {
	case OsLinux:
		return "Linux"
	case OsMacos:
		return "Macos"
	case OsWindows:
		return "Windows"
	default:
		return "Other"
	}
}

// CpuVendor is a best-effort CPU vendor classification.
type CpuVendor int

const (
	CpuVendorUnknown CpuVendor = iota
	CpuVendorIntel
	CpuVendorAmd
	CpuVendorArm
)

func (v CpuVendor) String() string {
	switch v {
	case CpuVendorIntel:
		return "Intel"
	case CpuVendorAmd:
		return "Amd"
	case CpuVendorArm:
		return "Arm"
	default:
		return "Unknown"
	}
}

// HardwareInventory is a snapshot of host hardware from cheap probes.
type HardwareInventory struct {
	OS                  OsKind
	CPUVendor           CpuVendor
	SIMD                simd.Backend
	PhysicalCores       int
	LogicalCores        int
	NumaNodes           int
	MinNodeRAMBytes     uint64
	TotalRAMBytes       uint64
	HasGPU              bool
	GPUFamily           *gpucluster.Family
	GPUVRAMBytes        uint64
	HasMetal            bool
	HasCUDA             bool
	HasROCm             bool
	HasRDMA             bool
	IsWSL               bool
	ContainerMemLimit   *uint64
	Hugepages2MiBAvail  bool
}

// Summary returns a one-line hardware summary.
func (h HardwareInventory) Summary() string {
	gpu := "gpu=none"
	if h.HasGPU {
		fam := "unknown"
		if h.GPUFamily != nil {
			fam = h.GPUFamily.Slug()
		}
		gpu = "gpu=" + fam + " vram=" + strconv.FormatUint(h.GPUVRAMBytes/(1024*1024), 10) + " MiB"
	}
	return strings.Join([]string{
		"os=" + h.OS.String(),
		"cpu=" + h.CPUVendor.String(),
		"simd=" + h.SIMD.String(),
		"cores=" + strconv.Itoa(h.PhysicalCores) + " (" + strconv.Itoa(h.LogicalCores) + "t)",
		"numa=" + strconv.Itoa(h.NumaNodes),
		"ram=" + strconv.FormatUint(h.TotalRAMBytes/(1<<30), 10) + " GiB",
		gpu,
		"metal=" + strconv.FormatBool(h.HasMetal),
		"cuda=" + strconv.FormatBool(h.HasCUDA),
		"wsl=" + strconv.FormatBool(h.IsWSL),
	}, " ")
}

// Detect runs all hardware probes and returns an inventory.
func Detect() HardwareInventory {
	osKind := detectOS()
	physical := runtime.NumCPU()
	if physical < 1 {
		physical = 1
	}
	logical := physical
	minNodeRAM := uint64(4) << 30
	totalRAM := detectTotalRAMBytes()
	if totalRAM == 0 {
		totalRAM = minNodeRAM
	}

	gpus := gpucluster.DetectGPUs()
	hasGPU := len(gpus) > 0
	var vram uint64
	var fam *gpucluster.Family
	for _, g := range gpus {
		vram += uint64(g.MemoryTotalMiB) * 1024 * 1024
		if g.FamilyKnown && fam == nil {
			f := g.Family
			fam = &f
		}
	}

	inv := HardwareInventory{
		OS:                 osKind,
		CPUVendor:          detectCPUVendor(),
		SIMD:               simd.Preferred(),
		PhysicalCores:      physical,
		LogicalCores:       logical,
		NumaNodes:          detectNumaNodes(),
		MinNodeRAMBytes:    minNodeRAM,
		TotalRAMBytes:      totalRAM,
		HasGPU:             hasGPU,
		GPUFamily:          fam,
		GPUVRAMBytes:       vram,
		HasMetal:           runtime.GOOS == "darwin",
		HasCUDA:            hasGPU,
		HasROCm:            false,
		HasRDMA:            false,
		IsWSL:              detectWSL(),
		ContainerMemLimit:  detectCgroupMemLimit(),
		Hugepages2MiBAvail: detectHugepages2MiB(),
	}
	return inv
}

func detectOS() OsKind {
	switch runtime.GOOS {
	case "linux":
		return OsLinux
	case "darwin":
		return OsMacos
	case "windows":
		return OsWindows
	default:
		return OsOther
	}
}

func detectTotalRAMBytes() uint64 {
	if runtime.GOOS != "linux" {
		return 0
	}
	data, err := os.ReadFile("/proc/meminfo")
	if err != nil {
		return 0
	}
	for _, line := range strings.Split(string(data), "\n") {
		if !strings.HasPrefix(line, "MemTotal:") {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}
		kb, err := strconv.ParseUint(fields[1], 10, 64)
		if err != nil {
			continue
		}
		return kb * 1024
	}
	return 0
}

func detectCPUVendor() CpuVendor {
	if runtime.GOARCH == "arm" || runtime.GOARCH == "arm64" {
		return CpuVendorArm
	}
	if runtime.GOOS != "linux" {
		return CpuVendorUnknown
	}
	data, err := os.ReadFile("/proc/cpuinfo")
	if err != nil {
		return CpuVendorUnknown
	}
	lower := strings.ToLower(string(data))
	switch {
	case strings.Contains(lower, "authenticamd"):
		return CpuVendorAmd
	case strings.Contains(lower, "genuineintel"):
		return CpuVendorIntel
	default:
		return CpuVendorUnknown
	}
}

func detectNumaNodes() int {
	if runtime.GOOS != "linux" {
		return 1
	}
	entries, err := os.ReadDir("/sys/devices/system/node")
	if err != nil {
		return 1
	}
	n := 0
	for _, e := range entries {
		if strings.HasPrefix(e.Name(), "node") {
			n++
		}
	}
	if n < 1 {
		return 1
	}
	return n
}

func detectWSL() bool {
	if runtime.GOOS != "linux" {
		return false
	}
	for _, path := range []string{"/proc/sys/kernel/osrelease", "/proc/version"} {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		lower := strings.ToLower(string(data))
		if strings.Contains(lower, "microsoft") || strings.Contains(lower, "wsl") {
			return true
		}
	}
	return false
}

func detectCgroupMemLimit() *uint64 {
	if runtime.GOOS != "linux" {
		return nil
	}
	if limit := readCgroupV2Limit("/sys/fs/cgroup/memory.max"); limit != nil {
		return limit
	}
	return readCgroupV1Limit("/sys/fs/cgroup/memory/memory.limit_in_bytes")
}

func readCgroupV2Limit(path string) *uint64 {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	trimmed := strings.TrimSpace(string(data))
	if trimmed == "max" || trimmed == "" {
		return nil
	}
	n, err := strconv.ParseUint(trimmed, 10, 64)
	if err != nil || n == 0 || n >= ^uint64(0) {
		return nil
	}
	return &n
}

func readCgroupV1Limit(path string) *uint64 {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	n, err := strconv.ParseUint(strings.TrimSpace(string(data)), 10, 64)
	if err != nil || n == 0 || n >= (1<<60) {
		return nil
	}
	return &n
}

func detectHugepages2MiB() bool {
	if runtime.GOOS != "linux" {
		return false
	}
	data, err := os.ReadFile("/sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages")
	if err != nil {
		return false
	}
	n, err := strconv.ParseUint(strings.TrimSpace(string(data)), 10, 64)
	return err == nil && n > 0
}

// IsSkylakeSP reports whether the host looks like Intel Skylake-SP (AVX-512 regression gate).
func IsSkylakeSP() bool {
	if runtime.GOOS != "linux" {
		return false
	}
	data, err := os.ReadFile("/proc/cpuinfo")
	if err != nil {
		return false
	}
	lower := strings.ToLower(string(data))
	return strings.Contains(lower, "skylake") && strings.Contains(lower, "xeon")
}

File: oxidize-python/oxidize_python/core/autotune/detect.py
"""Hardware detection for autotune (mirrors oxidize-golang/core/autotune/detect.go)."""

from __future__ import annotations

import os
import platform
import re
from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional

from oxidize_python.gpucluster import GpuFamily, DetectedGpu, detect_gpus
from oxidize_python.core.simd.simd import Backend, preferred


class OsKind(Enum):
    LINUX = auto()
    MACOS = auto()
    WINDOWS = auto()
    OTHER = auto()


class CpuVendor(Enum):
    UNKNOWN = auto()
    INTEL = auto()
    AMD = auto()
    ARM = auto()


@dataclass
class HardwareInventory:
    os: OsKind
    cpu_vendor: CpuVendor
    simd: Backend
    physical_cores: int
    logical_cores: int
    numa_nodes: int
    min_node_ram_bytes: int
    total_ram_bytes: int
    has_gpu: bool
    gpu_family: Optional[GpuFamily]
    gpu_vram_bytes: int
    has_metal: bool
    has_cuda: bool
    has_rocm: bool
    has_rdma: bool
    is_wsl: bool
    container_mem_limit: Optional[int]
    hugepages_2mib_avail: bool

    def summary(self) -> str:
        gpu = "gpu=none"
        if self.has_gpu:
            fam = self.gpu_family.name.lower() if self.gpu_family else "unknown"
            gpu = f"gpu={fam} vram={self.gpu_vram_bytes // (1024 * 1024)} MiB"
        return (
            f"os={self.os.name} cpu={self.cpu_vendor.name} simd={self.simd.name} "
            f"cores={self.physical_cores} ({self.logical_cores}t) numa={self.numa_nodes} "
            f"ram={self.total_ram_bytes // (1 << 30)} GiB {gpu} "
            f"metal={self.has_metal} cuda={self.has_cuda} wsl={self.is_wsl}"
        )


def detect() -> HardwareInventory:
    os_kind = _detect_os()
    physical = os.cpu_count() or 1
    logical = physical
    min_node = 4 << 30
    total = _detect_total_ram_bytes() or min_node

    gpus = detect_gpus()
    has_gpu = len(gpus) > 0
    vram = sum(int(g.memory_total_mib) * 1024 * 1024 for g in gpus)
    fam: Optional[GpuFamily] = None
    for g in gpus:
        if g.family is not None and fam is None:
            fam = g.family

    return HardwareInventory(
        os=os_kind,
        cpu_vendor=_detect_cpu_vendor(),
        simd=preferred(),
        physical_cores=physical,
        logical_cores=logical,
        numa_nodes=_detect_numa_nodes(),
        min_node_ram_bytes=min_node,
        total_ram_bytes=total,
        has_gpu=has_gpu,
        gpu_family=fam,
        gpu_vram_bytes=vram,
        has_metal=platform.system() == "Darwin",
        has_cuda=has_gpu,
        has_rocm=False,
        has_rdma=False,
        is_wsl=_detect_wsl(),
        container_mem_limit=_detect_cgroup_mem_limit(),
        hugepages_2mib_avail=_detect_hugepages_2mib(),
    )


def is_skylake_sp() -> bool:
    if platform.system() != "Linux":
        return False
    try:
        data = open("/proc/cpuinfo", encoding="utf-8").read().lower()
    except OSError:
        return False
    return "skylake" in data and "xeon" in data


def _detect_os() -> OsKind:
    system = platform.system()
    if system == "Linux":
        return OsKind.LINUX
    if system == "Darwin":
        return OsKind.MACOS
    if system == "Windows":
        return OsKind.WINDOWS
    return OsKind.OTHER


def _detect_total_ram_bytes() -> int:
    if platform.system() != "Linux":
        return 0
    try:
        with open("/proc/meminfo", encoding="utf-8") as f:
            for line in f:
                if line.startswith("MemTotal:"):
                    kb = int(line.split()[1])
                    return kb * 1024
    except OSError:
        return 0
    return 0


def _detect_cpu_vendor() -> CpuVendor:
    machine = platform.machine().lower()
    if machine.startswith("arm") or machine.startswith("aarch"):
        return CpuVendor.ARM
    if platform.system() != "Linux":
        return CpuVendor.UNKNOWN
    try:
        data = open("/proc/cpuinfo", encoding="utf-8").read().lower()
    except OSError:
        return CpuVendor.UNKNOWN
    if "authenticamd" in data:
        return CpuVendor.AMD
    if "genuineintel" in data:
        return CpuVendor.INTEL
    return CpuVendor.UNKNOWN


def _detect_numa_nodes() -> int:
    if platform.system() != "Linux":
        return 1
    try:
        nodes = [n for n in os.listdir("/sys/devices/system/node") if n.startswith("node")]
        return max(len(nodes), 1)
    except OSError:
        return 1


def _detect_wsl() -> bool:
    if platform.system() != "Linux":
        return False
    for path in ("/proc/sys/kernel/osrelease", "/proc/version"):
        try:
            data = open(path, encoding="utf-8").read().lower()
        except OSError:
            continue
        if "microsoft" in data or "wsl" in data:
            return True
    return False


def _detect_cgroup_mem_limit() -> Optional[int]:
    if platform.system() != "Linux":
        return None
    for path in ("/sys/fs/cgroup/memory.max", "/sys/fs/cgroup/memory/memory.limit_in_bytes"):
        try:
            raw = open(path, encoding="utf-8").read().strip()
        except OSError:
            continue
        if raw in ("", "max"):
            continue
        try:
            n = int(raw)
        except ValueError:
            continue
        if 0 < n < (1 << 60):
            return n
    return None


def _detect_hugepages_2mib() -> bool:
    path = "/sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages"
    try:
        n = int(open(path, encoding="utf-8").read().strip())
        return n > 0
    except (OSError, ValueError):
        return False
